### PR TITLE
[Feature](partition) Support insert overwrite auto detect partition

### PR DESCRIPTION
**Pull request by zclllyybb**
## Proposed changes

Issue Number: close #xxx

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: be/src/exec/tablet_info.cpp [hunk 1]
```diff
@@ -610,4 +615,68 @@ Status VOlapTablePartitionParam::add_partitions(
...
+        }
+        part->indexes = t_part.indexes;
+        std::sort(part->indexes.begin(), part->indexes.end(),
+                  [](const OlapTableIndexTablets& lhs, const OlapTableIndexTablets& rhs) {
+                      return lhs.index_id < rhs.index_id;
+                  });
+        // check index
+        for (int j = 0; j < num_indexes; ++j) {
+            if (part->indexes[j].index_id != _schema->indexes()[j]->index_id) {
+                std::stringstream ss;
```
Comment (HappenLee) [inline be/src/exec/tablet_info.cpp:654]: do not use `std::stringstream`

File: be/src/vec/sink/vrow_distribution.cpp [hunk 2]
```diff
@@ -109,6 +111,54 @@ Status VRowDistribution::automatic_create_partition() {
...
+            },
+            time_out));
+
+    Status status(Status::create(result.status));
+    LOG(WARNING) << "auto detect replace partition result: " << result;
+    if (result.status.status_code == TStatusCode::OK) {
+        RETURN_IF_ERROR(_vpartition->replace_partitions(_partitions, result.partitions));
+        // reuse the function as the args' structure are same. it add nodes/locations and incremental_open
+        auto result_as_create = cast_as_create_result(result);
+        RETURN_IF_ERROR(_create_partition_callback(_caller, &result_as_create));
```
Comment (HappenLee) [inline be/src/vec/sink/vrow_distribution.cpp:156]: where init `_create_partition_callback`, seems nullptr?

Comment (zclllyybb) [reply be/src/vec/sink/vrow_distribution.cpp:156]: in row_distribution's init. it reused auto_partition's same function.

**Comment by zclllyybb** [comment]
run buildall

File: be/src/exec/tablet_info.cpp [hunk 3]
```diff
@@ -40,13 +41,12 @@
...
 #include "runtime/raw_value.h"
 #include "runtime/types.h"
-#include "util/hash_util.hpp"
 #include "util/string_parser.hpp"
 #include "util/string_util.h"
 #include "vec/columns/column.h"
 #include "vec/columns/column_nullable.h"
-#include "vec/common/assert_cast.h"
-#include "vec/common/string_ref.h"
+// NOLINTNEXTLINE(unused-includes)
```
Comment (HappenLee) [inline be/src/exec/tablet_info.cpp:48]: remove unless comment

Comment (zclllyybb) [reply be/src/exec/tablet_info.cpp:48]: this is a lint for clangd to inhibit some wrong check

**Comment by zclllyybb** [comment]
run buildall

File: be/src/exec/tablet_info.h [hunk 4]
```diff
@@ -172,6 +172,12 @@ class VOlapTablePartitionParam {
                                       VOlapTablePartition*& partition) const {
         auto it = _is_in_partition ? _partitions_map->find(std::tuple {block, row, true})
                                    : _partitions_map->upper_bound(std::tuple {block, row, true});
+        // enable if need
```
Comment (HappenLee) [inline be/src/exec/tablet_info.h:175]: remove the unless comment

**Comment by zclllyybb** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/analysis/PartitionNames.java [hunk 5]
```diff
@@ -48,36 +48,36 @@ public class PartitionNames implements ParseNode, Writable {
     // true if these partitions are temp partitions
     @SerializedName(value = "isTemp")
     private final boolean isTemp;
-    private final boolean allPartitions;
+    private final boolean autoReplace;
```
Comment (Jibing-Li) [inline fe/fe-core/src/main/java/org/apache/doris/analysis/PartitionNames.java:51]: Why do we change the name of this variable? It's kind of misleading with the new name.

[the remark this answers is not in the document]
Comment (zclllyybb) [reply fe/fe-core/src/main/java/org/apache/doris/analysis/PartitionNames.java:51]: good idea. will fix it

File: fe/fe-core/src/main/java/org/apache/doris/catalog/OlapTable.java [hunk 6]
```diff
@@ -1106,6 +1106,12 @@ public Set<String> getPartitionNames() {
         return Sets.newHashSet(nameToPartition.keySet());
     }
 
+    public List<String> getPartitionNamesByIds(List<Long> partitionIds) {
+        return partitionIds.stream().map(id -> {
```
[the remark this answers is not in the document]
Comment (zclllyybb) [reply fe/fe-core/src/main/java/org/apache/doris/catalog/OlapTable.java:1110]: yes, it's guaranteed by caller. I'll change it name to reminder. thanks for checking

**Comment by zclllyybb** [comment]
run buildalll

**Comment by zclllyybb** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/insertoverwrite/InsertOverwriteManager.java [hunk 7]
```diff
@@ -69,12 +82,85 @@ public long registerTask(long dbId, long tableId, List<String> tempPartitionName
...
+    public ReentrantLock getLock(long groupId) {
+        return taskLocks.get(groupId);
+    }
+
+    public void taskGroupFail(long groupId) {
+        LOG.info("insert overwrite auto detect partition task group [" + groupId + "] failed");
+        for (Long taskId : taskGroups.get(groupId)) {
+            taskFail(taskId);
+        }
+        cleanTaskGroup(groupId);
```
Comment (zddr) [inline fe/fe-core/src/main/java/org/apache/doris/insertoverwrite/InsertOverwriteManager.java:140]: If fe is killed before `cleanTaskGroup`, will dirty data be generated?

Comment (zclllyybb) [reply fe/fe-core/src/main/java/org/apache/doris/insertoverwrite/InsertOverwriteManager.java:140]: no. taskSuccess or taskFail will deal with the transaction. cleanTaskGroup just clean some in-memory record

**Review by morrySnow — APPROVED** [review]

**Review by HappenLee — APPROVED** [review]
LGTM

**Pull request merged**